File: packages/sdk/src/discovery.ts
import { decodeProtectedHeader, importJWK, type JWK } from "jose";
import { assertPublicP256 } from "./crypto.js";
import { WeldallAuthError } from "./errors.js";

export type DiscoveryMetadata = { issuer: string; jwks_uri: string };

const jsonResponse = async (response: Response): Promise<unknown> => {
  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  const contentType = response.headers.get("content-type")?.split(";", 1)[0]?.trim().toLowerCase();
  if (contentType !== "application/json" && contentType !== "application/jwk-set+json")
    throw new Error("unexpected content type");
  const text = await response.text();
  if (text.length > 256_000) throw new Error("response too large");
  return JSON.parse(text);
};

const JWKS_TTL_MS = 60_000;

export class WeldallDiscovery {
  private metadata?: DiscoveryMetadata;
  private keys = new Map<string, JWK>();
  private metadataPromise: Promise<DiscoveryMetadata> | undefined;
  private jwksPromise: Promise<void> | undefined;
  private jwksExpiresAt = 0;
  private unknownRefreshAt = 0;

  constructor(
    private readonly host: string,
    private readonly timeoutMs: number,
  ) {}

  async ready(): Promise<void> {
    await this.refreshJwks(false);
  }

  async getKey(
    token: string,
    refreshForValidationFailure = false,
  ): Promise<{ kid: string; jwk: JWK }> {
    let kid: unknown;
    try {
      const header = decodeProtectedHeader(token);
      if (header.alg !== "ES256" || header.typ !== "oauth-id-jag+jwt" || header.crit)
        throw new Error("header");
      kid = header.kid;
    } catch {
      throw new WeldallAuthError("invalid_grant", "invalid ID-JAG header");
    }
    if (typeof kid !== "string" || !kid)
      throw new WeldallAuthError("invalid_grant", "invalid ID-JAG header");
    await this.refreshJwks(false);
    let jwk = this.keys.get(kid);
    if ((refreshForValidationFailure || !jwk) && Date.now() >= this.unknownRefreshAt) {
      this.unknownRefreshAt = Date.now() + 5_000;
      await this.refreshJwks(true);
      jwk = this.keys.get(kid);
    }
    if (!jwk) throw new WeldallAuthError("invalid_grant", "unknown Weldall signing key");
    return { kid, jwk };
  }

  private async discover(): Promise<DiscoveryMetadata> {
    if (this.metadata) return this.metadata;
    if (this.metadataPromise) return this.metadataPromise;
    this.metadataPromise = this.fetchMetadata().finally(() => {
      this.metadataPromise = undefined;
    });
    return this.metadataPromise;
  }

  private async fetchMetadata(): Promise<DiscoveryMetadata> {
    try {
      const value = await this.fetchJson(`${this.host}/.well-known/oauth-authorization-server`);
      if (!value || typeof value !== "object" || Array.isArray(value))
        throw new Error("invalid metadata");
      const { issuer, jwks_uri: jwksUri } = value as Record<string, unknown>;
      if (issuer !== this.host || typeof jwksUri !== "string") throw new Error("issuer mismatch");
      const url = new URL(jwksUri);
      if (url.origin !== this.host || url.username || url.password || url.search || url.hash)
        throw new Error("unsafe jwks_uri");
      this.metadata = { issuer, jwks_uri: url.toString() };
      return this.metadata;
    } catch {
      throw new WeldallAuthError("temporarily_unavailable", "Weldall discovery failed", 503);
    }
  }

  private async refreshJwks(force: boolean): Promise<void> {
    if (this.jwksPromise) return this.jwksPromise;
    if (!force && this.keys.size && Date.now() < this.jwksExpiresAt) return;
    this.jwksPromise = (async () => {
      const metadata = await this.discover();
      try {
        const value = await this.fetchJson(metadata.jwks_uri);
        if (!value || typeof value !== "object" || Array.isArray(value))
          throw new Error("invalid JWKS");
        const keys = (value as { keys?: unknown }).keys;
        if (!Array.isArray(keys) || keys.length < 1 || keys.length > 20)
          throw new Error("invalid JWKS size");
        const next = new Map<string, JWK>();
        for (const candidate of keys) {
          if (!candidate || typeof candidate !== "object" || Array.isArray(candidate))
            throw new Error("invalid JWK");
          const jwk = candidate as JWK;
          if (typeof jwk.kid !== "string" || !jwk.kid || next.has(jwk.kid))
            throw new Error("invalid JWK kid");
          await assertPublicP256(jwk);
          await importJWK(jwk, "ES256");
          next.set(jwk.kid, jwk);
        }
        this.keys = next;
        this.jwksExpiresAt = Date.now() + JWKS_TTL_MS;
      } catch (error) {
        if (error instanceof WeldallAuthError) throw error;
        throw new WeldallAuthError("temporarily_unavailable", "Weldall JWKS fetch failed", 503);
      }
    })().finally(() => {
      this.jwksPromise = undefined;
    });
    return this.jwksPromise;
  }

  private async fetchJson(url: string): Promise<unknown> {
    const signal = AbortSignal.timeout(this.timeoutMs);
    return jsonResponse(
      await fetch(url, { redirect: "error", signal, headers: { accept: "application/json" } }),
    );
  }
}

File: packages/python-sdk/src/weldall/discovery.py
"""Authorization-server metadata and rotating JWKS discovery."""

from __future__ import annotations

import json
import math
import threading
import time
from typing import Any

import httpx

from ._url import ParsedUrl, has_credentials, parse_absolute, serialize
from .crypto import assert_public_p256
from .errors import WeldallAuthError
from .jwt import decode_protected_header
from .types import JWK

_JWKS_TTL_SECONDS = 60.0
_UNKNOWN_REFRESH_SECONDS = 5.0
_MAX_BODY = 256_000


class WeldallDiscovery:
    """Synchronous, thread-safe discovery with shared in-flight results and failures."""

    def __init__(
        self,
        issuer: str,
        fetch_origin: str,
        timeout_ms: int = 5_000,
        *,
        http_client: httpx.Client | None = None,
    ) -> None:
        self.issuer = issuer
        self.fetch_origin = fetch_origin.rstrip("/")
        self.timeout_ms = timeout_ms
        self._client = http_client or httpx.Client(
            timeout=timeout_ms / 1000,
            follow_redirects=False,
        )
        self._owns_client = http_client is None
        self._metadata: dict[str, str] | None = None
        self._keys: dict[str, JWK] = {}
        self._jwks_expires_at = 0.0
        self._unknown_refresh_at = 0.0
        self._condition = threading.Condition()
        self._metadata_loading = False
        self._metadata_error: WeldallAuthError | None = None
        self._jwks_loading = False
        self._jwks_error: WeldallAuthError | None = None

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def ready(self) -> None:
        self._refresh_jwks(False)

    def get_key(self, token: str, refresh_for_validation_failure: bool = False) -> tuple[str, JWK]:
        return self.get_signing_key(token, "oauth-id-jag+jwt", refresh_for_validation_failure)

    def get_signing_key(
        self,
        token: str,
        expected_type: str,
        refresh_for_validation_failure: bool = False,
    ) -> tuple[str, JWK]:
        try:
            header = decode_protected_header(token)
            if (
                header.get("alg") != "ES256"
                or header.get("typ") != expected_type
                or "crit" in header
            ):
                raise ValueError("header")
            kid = header.get("kid")
        except Exception as error:
            raise WeldallAuthError("invalid_grant", "invalid signed assertion header") from error
        if not isinstance(kid, str) or not kid:
            raise WeldallAuthError("invalid_grant", "invalid signed assertion header")
        self._refresh_jwks(False)
        with self._condition:
            jwk = self._keys.get(kid)
        if refresh_for_validation_failure or jwk is None:
            self._refresh_jwks(True, throttled=True)
            with self._condition:
                jwk = self._keys.get(kid)
        if jwk is None:
            raise WeldallAuthError("invalid_grant", "unknown Weldall signing key")
        return kid, dict(jwk)

    def _discover(self) -> dict[str, str]:
        with self._condition:
            if self._metadata is not None:
                return dict(self._metadata)
            if self._metadata_loading:
                while self._metadata_loading:
                    self._condition.wait()
                if self._metadata is not None:
                    return dict(self._metadata)
                assert self._metadata_error is not None
                raise _copy_error(self._metadata_error)
            self._metadata_loading = True
            self._metadata_error = None

        metadata: dict[str, str] | None = None
        failure: WeldallAuthError | None = None
        try:
            value = self._fetch_json(f"{self.fetch_origin}/.well-known/oauth-authorization-server")
            if not isinstance(value, dict):
                raise ValueError("invalid metadata")
            issuer = value.get("issuer")
            jwks_uri = value.get("jwks_uri")
            if issuer != self.issuer or not isinstance(jwks_uri, str):
                raise ValueError("issuer mismatch")
            parsed = parse_absolute(jwks_uri, "jwks_uri")
            canonical = serialize(parsed)
            if (
                _origin(parsed) != self.issuer
                or has_credentials(parsed)
                or bool(parsed.query)
                or bool(parsed.fragment)
            ):
                raise ValueError("unsafe jwks_uri")
            metadata = {"issuer": issuer, "jwks_uri": canonical}
        except Exception as error:
            failure = WeldallAuthError("temporarily_unavailable", "Weldall discovery failed", 503)
            failure.__cause__ = error
        finally:
            with self._condition:
                if metadata is not None:
                    self._metadata = metadata
                self._metadata_error = failure
                self._metadata_loading = False
                self._condition.notify_all()
        if failure is not None:
            raise failure
        assert metadata is not None
        return dict(metadata)

    def _refresh_jwks(self, force: bool, *, throttled: bool = False) -> None:
        with self._condition:
            if self._jwks_loading:
                while self._jwks_loading:
                    self._condition.wait()
                if self._jwks_error is not None:
                    raise _copy_error(self._jwks_error)
                return
            now = time.monotonic()
            if not force and self._keys and now < self._jwks_expires_at:
                return
            if throttled:
                if now < self._unknown_refresh_at:
                    return
                self._unknown_refresh_at = now + _UNKNOWN_REFRESH_SECONDS
            self._jwks_loading = True
            self._jwks_error = None

        next_keys: dict[str, JWK] | None = None
        failure: WeldallAuthError | None = None
        try:
            metadata = self._discover()
            parsed = parse_absolute(metadata["jwks_uri"], "jwks_uri")
            value = self._fetch_json(f"{self.fetch_origin}{parsed.path}")
            if not isinstance(value, dict):
                raise ValueError("invalid JWKS")
            keys = value.get("keys")
            if not isinstance(keys, list) or not 1 <= len(keys) <= 20:
                raise ValueError("invalid JWKS size")
            next_keys = {}
            for candidate in keys:
                if not isinstance(candidate, dict):
                    raise ValueError("invalid JWK")
                kid = candidate.get("kid")
                if not isinstance(kid, str) or not kid or kid in next_keys:
                    raise ValueError("invalid JWK kid")
                assert_public_p256(candidate)
                next_keys[kid] = dict(candidate)
        except WeldallAuthError as error:
            failure = error
        except Exception as error:
            failure = WeldallAuthError("temporarily_unavailable", "Weldall JWKS fetch failed", 503)
            failure.__cause__ = error
        finally:
            with self._condition:
                if next_keys is not None:
                    self._keys = next_keys
                    self._jwks_expires_at = time.monotonic() + _JWKS_TTL_SECONDS
                self._jwks_error = failure
                self._jwks_loading = False
                self._condition.notify_all()
        if failure is not None:
            raise failure

    def _fetch_json(self, url: str) -> Any:
        with self._client.stream("GET", url, headers={"accept": "application/json"}) as response:
            response.raise_for_status()
            content_type = response.headers.get("content-type", "").split(";", 1)[0].strip().lower()
            if content_type not in {"application/json", "application/jwk-set+json"}:
                raise ValueError("unexpected content type")
            declared = response.headers.get("content-length")
            if declared is not None:
                try:
                    length = float(declared)
                except ValueError:
                    length = math.nan
                if math.isfinite(length) and length > _MAX_BODY:
                    raise ValueError("response too large")
            chunks: list[bytes] = []
            size = 0
            for chunk in response.iter_bytes():
                size += len(chunk)
                if size > _MAX_BODY:
                    raise ValueError("response too large")
                chunks.append(chunk)
        return json.loads(b"".join(chunks).decode("utf-8", errors="strict"))


def _origin(parsed: ParsedUrl) -> str:
    return serialize(parsed, path="", query="", fragment="").rstrip("/")


def _copy_error(error: WeldallAuthError) -> WeldallAuthError:
    return WeldallAuthError(
        error.code,
        error.message,
        error.status,
        error.required_scopes,
        error.reason,
    )
